File: nexustrace-frontend/types/rag.ts
export type RagProvider = "auto" | "openai" | "gemini";

export interface RagQuery {
  question: string;
  case_id: string;
  chat_history?: ChatHistoryMessage[];
  provider?: RagProvider;
}

export interface ChatHistoryMessage {
  role: "user" | "assistant";
  content: string;
}

export interface RagChunk {
  chunk_id: string;
  content: string;
  source: string;
  similarity_score: number;
}

export interface SourceAttribution {
  filename: string;
  evidence_id: string;
  file_type: string;
  pages_referenced: number[];
}

export interface RagResponse {
  query_id: string;
  answer: string;
  cited_chunks: string[];
  reasoning_summary: string;
  confidence_score: number;
  sources: SourceAttribution[];
  provider_requested?: RagProvider;
  provider_used?: string;
}

export interface RagExplanation {
  query_id: string;
  retrieved_chunks: RagChunk[];
  graph_path: string[];
  reasoning: string;
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  query_id?: string;
  cited_chunks?: string[];
  sources?: SourceAttribution[];
  confidence?: number;
  provider_used?: string;
  timestamp: Date;
}

export interface FeedbackRequest {
  query_id: string;
  is_correct: boolean;
  comment?: string;
}

export interface QueryHistory {
  query_id: string;
  case_id: string;
  question: string;
  answer: string | null;
  timestamp: number;
  chunks_retrieved: number;
}

File: nexustrace-backend/app/schemas/rag.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field

class ChatHistoryMessage(BaseModel):
    role: str  # "user" or "assistant" 
    content: str

class RAGQuery(BaseModel):
    case_id: str
    question: str
    chat_history: Optional[List[ChatHistoryMessage]] = None
    provider: Literal["auto", "openai", "gemini"] = "auto"

class SourceAttribution(BaseModel):
    filename: str
    evidence_id: str = ""
    file_type: str = ""
    pages_referenced: List[int] = Field(default_factory=list)

class RAGResponse(BaseModel):
    query_id: str
    answer: str
    cited_chunks: List[str]
    reasoning_summary: str
    confidence_score: float
    sources: List[SourceAttribution] = Field(default_factory=list)
    provider_requested: Literal["auto", "openai", "gemini"] = "auto"
    provider_used: str = "unknown"

class ExplanationResponse(BaseModel):
    query_id: str
    retrieved_chunks: List[Dict[str, Any]] = Field(default_factory=list)
    graph_path: List[str] = Field(default_factory=list)
    reasoning: str = ""
    # Legacy fields kept for backward compatibility with older clients.
    question: Optional[str] = None
    graph_expansion: List[Dict[str, Any]] = Field(default_factory=list)

class QueryHistory(BaseModel):
    query_id: str
    case_id: str
    question: str
    answer: Optional[str] = None
    timestamp: int
    chunks_retrieved: Optional[int] = 0
